File: Data Structures/Current Dump/2/2/03.LinearDataStructuresStacksAndQueues/StacksAndQueuesHomework/05.LinkedStack/LinkedStack.cs
namespace _05.LinkedStack
{
    using System;

    public class LinkedStack<T>
    {
        private Node<T> head;

        public int Count { get; private set; }

        public void Push(T element)
        {
            this.head = new Node<T>(element, this.head);

            this.Count++;
        }

        public T Pop()
        {
            if (this.Count == 0)
            {
                throw new InvalidOperationException("The stack is empty!");
            }

            var element = this.head.Value;
            this.head = this.head.Next;
            this.Count--;

            return element;
        }

        public T[] ToArray()
        {
            var array = new T[this.Count];
            var currNode = this.head;
            for (int i = 0; i < this.Count; i++)
            {
                array[i] = currNode.Value;
                currNode = currNode.Next;
            }

            return array;
        }
    }
}

File: Java OOP Advanced/02. Generics/Demos/02. Java-OOP-Advanced-Generics-Demos/src/d1/genericClasses/LinkedStack.java
package d1.genericClasses;

public class LinkedStack<T> {

    private class Node {
        T item;
        Node next;

        Node() {
            this.item = null;
            this.next = null;
        }

        Node(T item, Node next) {
            this.item = item;
            this.next = next;
        }

        boolean end() {
            return this.item == null && this.next == null;
        }
    }

    private Node top = new Node();

    public void push(T item) {
        this.top = new Node(item, this.top);
    }

    public T pop() {
        T result = this.top.item;
        if (!this.top.end()) {
            this.top = this.top.next;
        }

        return result;
    }
}
